File: app/src/main/java/org/autojs/autojs/ui/common/OperationDialogBuilder.java
package org.autojs.autojs.ui.common;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.afollestad.materialdialogs.MaterialDialog;

import org.autojs.autojs.R;
import org.autojs.autojs.databinding.OperationDialogItemBinding;

import java.util.ArrayList;

/**
 * Created by Stardust on 2017/6/26.
 */

public class OperationDialogBuilder extends MaterialDialog.Builder {

    private final RecyclerView mOperations;
    private final ArrayList<Integer> mIds = new ArrayList<>();
    private final ArrayList<Integer> mIcons = new ArrayList<>();
    private final ArrayList<String> mTexts = new ArrayList<>();
    private View.OnClickListener mItemClickListener;

    public OperationDialogBuilder(@NonNull Context context) {
        super(context);
        mOperations = new RecyclerView(context);
        mOperations.setLayoutManager(new LinearLayoutManager(context));
        mOperations.setAdapter(new RecyclerView.Adapter<ViewHolder>() {
            @Override
            public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
                return new ViewHolder(LayoutInflater.from(parent.getContext()).inflate(R.layout.operation_dialog_item, parent, false));
            }

            @Override
            public void onBindViewHolder(ViewHolder holder, int position) {
                holder.itemView.setId(mIds.get(position));
                holder.binding.text.setText(mTexts.get(position));
                holder.binding.icon.setImageResource(mIcons.get(position));
                if (mItemClickListener != null) {
                    holder.itemView.setOnClickListener(mItemClickListener);
                }
            }

            @Override
            public int getItemCount() {
                return mIds.size();
            }
        });
        customView(mOperations, false);
    }

    public OperationDialogBuilder item(int id, int iconRes, int textRes) {
        return item(id, iconRes, getContext().getString(textRes));
    }

    public OperationDialogBuilder item(int id, int iconRes, String text) {
        mIds.add(id);
        mIcons.add(iconRes);
        mTexts.add(text);
        return this;
    }

    public OperationDialogBuilder bindItemClick(View.OnClickListener listener) {
        mItemClickListener = listener;
        return this;
    }

    static class ViewHolder extends RecyclerView.ViewHolder {
        final OperationDialogItemBinding binding;

        ViewHolder(View itemView) {
            super(itemView);
            binding = OperationDialogItemBinding.bind(itemView);
        }
    }
}

File: app/src/main/java/org/autojs/autojs/ui/common/OperationDialogBuilder.kt
package org.autojs.autojs.ui.common

import android.content.Context
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.afollestad.materialdialogs.MaterialDialog
import org.autojs.autojs.databinding.OperationDialogItemBinding
import butterknife.ButterKnife

class OperationDialogBuilder(context: Context) : MaterialDialog.Builder(context) {

    private val mIds = ArrayList<Int>()
    private val mIcons = ArrayList<Int>()
    private val mTexts = ArrayList<String>()
    private var mOnItemClickTarget: Any? = null

    init {
        val operations = RecyclerView(context).apply {
            layoutManager = LinearLayoutManager(context)
            adapter = object : RecyclerView.Adapter<ViewHolder>() {
                override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
                    val binding = OperationDialogItemBinding.inflate(
                        LayoutInflater.from(parent.context), parent, false
                    )
                    return ViewHolder(binding)
                }

                override fun onBindViewHolder(holder: ViewHolder, position: Int) {
                    holder.itemView.id = mIds[position]
                    holder.binding.text.text = mTexts[position]
                    holder.binding.icon.setImageResource(mIcons[position])
                    mOnItemClickTarget?.let { ButterKnife.bind(it, holder.itemView) }
                }

                override fun getItemCount(): Int = mIds.size
            }
        }
        customView(operations, false)
    }

    fun item(id: Int, iconRes: Int, textRes: Int): OperationDialogBuilder {
        return item(id, iconRes, context.getString(textRes))
    }

    fun item(id: Int, iconRes: Int, text: String): OperationDialogBuilder {
        mIds.add(id)
        mIcons.add(iconRes)
        mTexts.add(text)
        return this
    }

    fun bindItemClick(target: Any?): OperationDialogBuilder {
        mOnItemClickTarget = target
        return this
    }

    private class ViewHolder(val binding: OperationDialogItemBinding) : RecyclerView.ViewHolder(binding.root)
}
